File: src/main/java/org/alpacaindustries/iremiaminigamecore/minigame/MinigameBuilder.java
package org.alpacaindustries.iremiaminigamecore.minigame;

import org.bukkit.Location;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

/**
 * Builder pattern for configuring minigames
 */
public class MinigameBuilder {
    private int minPlayers = 1;
    private int maxPlayers = 16;
    private boolean allowJoinDuringGame = false;
    private Optional<Location> spawnPoint = Optional.empty();

    /**
     * Set the minimum number of players.
     */
    public @NotNull MinigameBuilder minPlayers(int minPlayers) {
        this.minPlayers = minPlayers;
        return this;
    }

    /**
     * Set the maximum number of players.
     */
    public @NotNull MinigameBuilder maxPlayers(int maxPlayers) {
        this.maxPlayers = maxPlayers;
        return this;
    }

    /**
     * Set whether players can join during the game.
     */
    public @NotNull MinigameBuilder allowJoinDuringGame(boolean allow) {
        this.allowJoinDuringGame = allow;
        return this;
    }

    /**
     * Set the spawn point for the minigame.
     */
    public @NotNull MinigameBuilder spawnPoint(@Nullable Location spawnPoint) {
        this.spawnPoint = Optional.ofNullable(spawnPoint);
        return this;
    }

    /**
     * Apply the builder's configuration to a Minigame instance.
     */
    public void applyTo(@NotNull Minigame minigame) {
        minigame.setMinPlayers(minPlayers);
        minigame.setMaxPlayers(maxPlayers);
        minigame.setAllowJoinDuringGame(allowJoinDuringGame);
        minigame.setSpawnPoint(spawnPoint.orElse(null));
    }
}

File: src/main/kotlin/org/alpacaindustries/iremiaminigamecore/minigame/MinigameBuilder.kt
package org.alpacaindustries.iremiaminigamecore.minigame

import org.bukkit.Location

/**
 * Builder pattern for configuring minigames
 */
class MinigameBuilder {
    private var minPlayers: Int = 1
    private var maxPlayers: Int = 16
    private var allowJoinDuringGame: Boolean = false
    private var spawnPoint: Location? = null

    /**
     * Set the minimum number of players.
     */
    fun minPlayers(minPlayers: Int): MinigameBuilder = apply {
        this.minPlayers = minPlayers
    }

    /**
     * Set the maximum number of players.
     */
    fun maxPlayers(maxPlayers: Int): MinigameBuilder = apply {
        this.maxPlayers = maxPlayers
    }

    /**
     * Set whether players can join during the game.
     */
    fun allowJoinDuringGame(allow: Boolean): MinigameBuilder = apply {
        this.allowJoinDuringGame = allow
    }

    /**
     * Set the spawn point for the minigame.
     */
    fun spawnPoint(spawnPoint: Location?): MinigameBuilder = apply {
        this.spawnPoint = spawnPoint
    }

    /**
     * Apply the builder's configuration to a Kotlin MinigameKt instance.
     */
    fun applyTo(minigame: Minigame) {
        minigame.minPlayers = minPlayers
        minigame.maxPlayers = maxPlayers
        minigame.isAllowJoinDuringGame = allowJoinDuringGame
        minigame.spawnPoint = spawnPoint
    }
}
